providers/implementations/keymgmt/mlkem_kmgmt.c

```diff
@@ -72,435 +40,441 @@ static OSSL_FUNC_keymgmt_import_types_fn mlkem_imexport_types;
-        /*
-         * ML-KEM keys always have all the parameters they need (i.e. none).
-         * Therefore we always return with 1, if asked about parameters.
-         */
-        ok = 1;
-
-        if ((selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0)
-            ok = ok && key->encoded_pubkey != NULL;
-
-        if ((selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0)
-            ok = ok && key->encoded_privkey != NULL;
-    }
-    debug_print("MLKEMKM has result %d\n", ok);
-    return ok;
+    /* Free all key material, zeroing any private parts. */
+    OPENSSL_free(key->pubkey);
+    ossl_mlkem_vcleanse_prvkey(v, &key->prvkey);
+    ossl_mlkem_ctx_free(key->ctx);
+    OPENSSL_free(key);
 }
 
-static int mlkem_match(const void *keydata1, const void *keydata2, int selection)
+static int mlkem_has(const void *vkey, int selection)
 {
-    const MLKEM768_KEY *key1 = keydata1;
-    const MLKEM768_KEY *key2 = keydata2;
-    int ok = 1;
+    static int selectable =
+        OSSL_KEYMGMT_SELECT_PUBLIC_KEY |
+        OSSL_KEYMGMT_SELECT_PRIVATE_KEY;
+    const MLKEM_KEY *key = vkey;
 
-    debug_print("MLKEMKM matching %p and %p\n", key1, key2);
-    if (!ossl_prov_is_running())
+    if (!ossl_prov_is_running() || key == NULL)
         return 0;
+    if ((selection & selectable) == 0)
+        return 1;
 
-    if ((selection & OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS) != 0)
-        ok = ok && key1->keytype == key2->keytype;
-
-    /* TODO(ML-KEM) */
-    debug_print("MLKEMKM matching for now NOT YET IMPLEMENTED\n");
-
-/* TODO(ML-KEM) template code to be completed as and when needed: */
-#ifdef UNDEF
-    if ((selection & OSSL_KEYMGMT_SELECT_KEYPAIR) != 0) {
-        int key_checked = 0;
-
-        if ((selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0) {
-            const uint8_t *pa = key1->pubkey;
-            const uint8_t *pb = key2->pubkey;
-
-            if (pa != NULL && pb != NULL) {
-                ok = ok
-                    && key1->keytype == key2->keytype
-                    && CRYPTO_memcmp(pa, pb, MLKEM768_PUBLICKEYBYTES) == 0;
-                key_checked = 1;
-            }
-        }
-        if (!key_checked
-            && (selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0) {
-            const uint8_t *pa = key1->privkey;
-            const uint8_t *pb = key2->privkey;
-
-            if (pa != NULL && pb != NULL) {
-                ok = ok
-                    && key1->keytype == key2->keytype
-                    && CRYPTO_memcmp(pa, pb, MLKEM768_SECRETKEYBYTES) == 0;
-                key_checked = 1;
-            }
-        }
-        ok = ok && key_checked;
+    /*
+     * Fail if the public key (or if requested the private key) is unavailable.
+     * The public key is available when either was provided.
+     */
+    if (!have_keys(key)
+        || ((selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0
+            && key->prvkey == NULL)) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_KEY);
```

Comment:
IMO the has function is asking whether a key has certain elements. It is not a "failure" if it does not, i.e. the "has" function successfully indicates that the key does not have the queried elements. So `ERR_raise` does not seem appropriate here to me.

Reply:
So it isn't the zero return value that's the problem, but rather that is expected to ask whether key material is there, even on not yet fully initialised keys.  So sure, dropping the "raise".